core/src/test/java/greencity/controller/HabitControllerTest.java

```diff
@@ -0,0 +1,290 @@
+import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.csrf;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
+
+import greencity.GreenCityApplication;
+import greencity.dto.PageableDto;
+import greencity.dto.habit.AddCustomHabitDtoRequest;
+import greencity.dto.habit.AddCustomHabitDtoResponse;
+import greencity.dto.habit.HabitDto;
+import greencity.dto.shoppinglistitem.ShoppingListItemDto;
+import greencity.dto.user.UserProfilePictureDto;
+import greencity.dto.user.UserVO;
+import greencity.exception.exceptions.BadRequestException;
+import greencity.service.HabitService;
+import greencity.service.LanguageService;
+import greencity.service.TagsService;
+import greencity.service.UserService;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.modelmapper.ModelMapper;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.data.domain.Pageable;
+import org.springframework.http.MediaType;
+import org.springframework.mock.web.MockMultipartFile;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.authority.SimpleGrantedAuthority;
+import org.springframework.security.test.context.support.WithMockUser;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.web.multipart.MultipartFile;
+import org.testcontainers.shaded.com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.util.*;
+
+@WithMockUser(username = "user")
+@WebMvcTest(HabitController.class)
+@ContextConfiguration(classes = {GreenCityApplication.class})
+class HabitControllerTest {
+
+    @Autowired
+    private MockMvc mockMvc;
+
+    @MockBean
+    private ModelMapper modelMapper;
+
+    @MockBean
+    private LanguageService languageService;
+
+    @MockBean
+    private UserService userService;
+
+    @MockBean
+    private TagsService tagsService;
+
+    @MockBean
+    private HabitService habitService;
+
+    private UserVO userVO;
+
+    @BeforeEach
+    void setUp() {
+        userVO = UserVO.builder()
+                .id(1L)
+                .name("John Doe")
+                .email("john@example.com")
+                .build();
+
+        when(userService.findByEmail(anyString())).thenReturn(userVO);
+        when(languageService.findAllLanguageCodes()).thenReturn(List.of("en"));
+    }
+
+    @Test
+    void getHabitById_HabitExists_ReturnsHabit() throws Exception {
+        HabitDto habitDto = HabitDto.builder()
```

Comment:
might be better to add new test object to ModelUtils

Reply:
Advantages
Centralization: Having a single location for model creation methods can make it easier to manage and update test data. This is particularly useful when multiple test classes utilize the same or similar data setups.
Reusability: It promotes reuse of code across different tests, reducing duplication and potentially decreasing the effort needed to write new tests.
Consistency: Ensures that all tests are using the same initial setup, which can improve the consistency of test outcomes.
Disadvantages
Coupling: Test classes become tightly coupled to the utility class, which might be problematic if the data models change frequently, as every change could impact numerous tests.
Complexity: For someone reviewing the code or running the tests, it might be difficult to trace back to the model creation logic if it's not directly within the test class. This could affect readability and maintainability.
Flexibility: It might limit the flexibility to have test-specific data variations if all tests are drawing from the same method for model setup.
Alternatives
If the disadvantages outweigh the benefits in your context, you might consider a few alternatives:

Factory Pattern: Implement a factory pattern specifically for test data generation. This can encapsulate the creation logic but still allow for flexibility by parametrizing the factory methods to produce different variants of the objects based on the needs of different tests.
Builder Pattern: Utilize a builder pattern within each test or a centralized class. This can provide a clear and flexible way of constructing complex objects step-by-step, and it’s especially useful when dealing with objects that require numerous parameters for construction.
Direct Setup in Tests: Sometimes, if tests vary greatly or readability is a major concern, it might be best to set up test data directly within the test classes, even if this leads to some duplication. This approach improves the readability of tests, as all setup details are immediately visible within each test class.